control/controlclient/auto.go

```diff
@@ -46,6 +47,79 @@ func (g *LoginGoal) sendLogoutError(err error) {
 
 var _ Client = (*Auto)(nil)
 
+// waitUnpause waits until the client is unpaused then returns. It only
+// returns an error if the client is closed.
+func (c *Auto) waitUnpause(routineLogName string) error {
+	c.mu.Lock()
+	if !c.paused {
+		c.mu.Unlock()
+		return nil
+	}
+	unpaused := c.unpausedChanLocked()
+	c.mu.Unlock()
+	c.logf("%s: awaiting unpause", routineLogName)
+	select {
+	case <-unpaused:
+		c.logf("%s: unpaused", routineLogName)
+		return nil
+	case <-c.quit:
+		return errors.New("quit")
+	}
+}
+
+// informRoutine is responsible for informing the server of worthy changes to
+// our local state. It runs in its own goroutine.
+func (c *Auto) informRoutine() {
+	defer close(c.informDone)
+	bo := backoff.NewBackoff("informRoutine", c.logf, 30*time.Second)
+	for {
+		if err := c.waitUnpause("informRoutine"); err != nil {
+			c.logf("informRoutine: exiting")
+			return
+		}
+		c.mu.Lock()
+		at := c.lastUpdateAt
+		ctx := c.mapCtx
+		needUpdate := at != c.lastUpdateInformed && !at.IsZero()
+		c.mu.Unlock()
+
```

Comment:
how about a new type for this informRoutine and breaking up this up into methods?